### PR TITLE
Add unit tests to verify yarn launcher config construction

### PR DESCRIPTION
**Pull request by ashvina**
This PR is related to https://github.com/twitter/heron/pull/1454#discussion_r81189009
cc @billonahill

### CODE REVIEW

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    expectedMap.put(Role.class.getSimpleName(), "role");
+    testConfigMap.put(Keys.environ(), "env");
+    expectedMap.put(Environ.class.getSimpleName(), "env");
+    testConfigMap.put(YarnContext.HERON_SCHEDULER_YARN_QUEUE, "queue");
+    expectedMap.put(JobQueue.class.getSimpleName(), "queue");
+    testConfigMap.put(YarnContext.YARN_SCHEDULER_DRIVER_MEMORY_MB, "100");
+    expectedMap.put(DriverMemory.class.getSimpleName(), "100");
+
+    Config.Builder builder = new Config.Builder();
+    for (String s : testConfigMap.keySet()) {
```
Comment (billonahill) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: please rename 's' to something like configKey or something descriprive

**Comment by billonahill** [comment]
Two general observations:

1. Do you have a concept of default config settings? If so can we have two tests, one for testing defaults are as expected and one for testing non-defaults?
2. How could we do this in a way where if someone adds a new config but forgets to update this test, it fails? Is there a way to require all config keys to be part of an enum or something like that, where adding one to the config/enum, but not a test verification for it would fail?

**Comment by ashvina** [comment]
I think I can address the second observation. 

Regarding observation 1, most of the configurations provided to REEF are heron configs and are not optional. For e.g. there is no default value of a topology name. These are provided to launcher by heron config loader which is independently unit tested. I am not sure if this kind of test is applicable here?

**Comment by ashvina** [comment]
Updated the unit test. It will fail if unexpected config is found.
cc @billonahill

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    for (NamedParameterNode<?> reefConfigNode : constructedConfig.getNamedParameters()) {
+      String value = expectedMap.get(reefConfigNode.getName());
+      if (value != null) {
+        Assert.assertEquals(value, constructedConfig.getNamedParameter(reefConfigNode));
+        expectedMap.remove(reefConfigNode.getName());
+      } else {
+        reefSpecificConfigs.remove(reefConfigNode.getName());
+      }
+    }
+    Assert.assertEquals(0, expectedMap.size());
```
Comment (billonahill) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: If either of these assertions fails, the reason won't be clear to the reader of the logs. Would you please add a string reason for the failure (e.g., an unexpected config found) and output the map/set.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    setConfigs(Keys.topologyPackageFile(), "package", TopologyPackageName.class);
+    setConfigs(Keys.cluster(), "cluster", Cluster.class);
+    setConfigs(Keys.role(), "role", Role.class);
+    setConfigs(Keys.environ(), "env", Environ.class);
+    setConfigs(YarnContext.HERON_SCHEDULER_YARN_QUEUE, "queue", JobQueue.class);
+    setConfigs(YarnContext.YARN_SCHEDULER_DRIVER_MEMORY_MB, "123", DriverMemory.class);
+    setConfigs(Keys.corePackageUri(), new File(".").getName(), HeronCorePackageName.class);
+
+    Set<String> reefSpecificConfigs = new HashSet<>();
+    reefSpecificConfigs.add("JobControlHandler");
```
Comment (billonahill) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Can these key values be static strings referenced from somewhere?

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    reefSpecificConfigs.add("NodeDescriptorHandler");
+    reefSpecificConfigs.add("ResourceAllocationHandler");
+    reefSpecificConfigs.add("ResourceStatusHandler");
+    reefSpecificConfigs.add("RuntimeStatusHandler");
+    reefSpecificConfigs.add("VerboseLogMode");
+    reefSpecificConfigs.add("HttpPort");
+    reefSpecificConfigs.add("DriverIdentifier");
+
+    Config.Builder builder = new Config.Builder();
+    for (String s : testConfigMap.keySet()) {
```
Comment (billonahill) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: please use a more descriptive field name than `s`.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    builder.put(Keys.stateManagerClass(), "statemanager");
+    builder.put(Keys.packingClass(), "packing");
+    Config config = builder.build();
+
+    spyLauncher.initialize(config, null);
+    Configuration constructedConfig = spyLauncher.getHMDriverConf();
+
+    for (NamedParameterNode<?> reefConfigNode : constructedConfig.getNamedParameters()) {
+      String value = expectedMap.get(reefConfigNode.getName());
+      if (value != null) {
```
Comment (billonahill) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: There are enough Map, Set, Config, Builder objects all referring to different expected/found/reef values that reading the test can get tricky to a reader without details of the config code implementation. Would you please add a few comments along the way helping explain things.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java (hunk 1 above)
Comment (ashvina) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Fixed

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java (hunk 2 above)
Comment (ashvina) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Fixed

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java (hunk 3 above)
Comment (ashvina) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Fixed

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java (hunk 4 above)
Comment (ashvina) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Fixed

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java (hunk 5 above)
Comment (ashvina) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/yarn/YarnLauncherTest.java]: Fixed

**Pull request merged**